conformance/echoserver/echoserver.go

```diff
@@ -0,0 +1,287 @@
+	NegotiatedProtocol string   `json:"negotiatedProtocol,omitempty"`
+	CipherSuite        string   `json:"cipherSuite"`
+}
+
+type preserveSlashes struct {
+	mux http.Handler
+}
+
+func (s *preserveSlashes) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	r.URL.Path = strings.Replace(r.URL.Path, "//", "/", -1)
+	s.mux.ServeHTTP(w, r)
+}
+
+// Context contains information about the context where the echoserver is running
+type Context struct {
+	Namespace string `json:"namespace"`
+	Ingress   string `json:"ingress"`
+	Service   string `json:"service"`
+	Pod       string `json:"pod"`
+}
+
+var context Context
+
+func main() {
+	httpPort := os.Getenv("HTTP_PORT")
+	if httpPort == "" {
+		httpPort = "3000"
+	}
+
+	httpsPort := os.Getenv("HTTPS_PORT")
+	if httpsPort == "" {
+		httpsPort = "8443"
+	}
+
+	context = Context{
+		Namespace: os.Getenv("NAMESPACE"),
+		Ingress:   os.Getenv("INGRESS_NAME"),
+		Service:   os.Getenv("SERVICE_NAME"),
+		Pod:       os.Getenv("POD_NAME"),
+	}
+
+	httpMux := http.NewServeMux()
+	httpMux.HandleFunc("/health", healthHandler)
+	httpMux.HandleFunc("/status/", statusHandler)
+	httpMux.HandleFunc("/", echoHandler)
+	httpHandler := &preserveSlashes{httpMux}
+
+	errchan := make(chan error)
+
+	go func() {
+		fmt.Printf("Starting server, listening on port %s (http)\n", httpPort)
+		err := http.ListenAndServe(fmt.Sprintf(":%s", httpPort), httpHandler)
+		if err != nil {
+			errchan <- err
+		}
+	}()
+
+	// Enable HTTPS if certificate and private key are given.
+	if os.Getenv("TLS_SERVER_CERT") != "" && os.Getenv("TLS_SERVER_PRIVKEY") != "" {
+		go func() {
+			fmt.Printf("Starting server, listening on port %s (https)\n", httpsPort)
+			err := listenAndServeTLS(fmt.Sprintf(":%s", httpsPort), os.Getenv("TLS_SERVER_CERT"), os.Getenv("TLS_SERVER_PRIVKEY"), os.Getenv("TLS_CLIENT_CACERTS"), httpHandler)
+			if err != nil {
+				errchan <- err
+			}
+		}()
+	}
+
+	select {
+	case err := <-errchan:
+		panic(fmt.Sprintf("Failed to start listening: %s\n", err.Error()))
+	}
+}
+
+func healthHandler(w http.ResponseWriter, r *http.Request) {
+	w.WriteHeader(200)
+	w.Write([]byte(`OK`))
+}
+
+func statusHandler(w http.ResponseWriter, r *http.Request) {
+	code := http.StatusBadRequest
+
+	re := regexp.MustCompile(`^/status/(\d\d\d)$`)
+	match := re.FindStringSubmatch(r.RequestURI)
+	if match != nil {
+		code, _ = strconv.Atoi(match[1])
+	}
+
+	w.WriteHeader(code)
+}
+
+func delayResponse(request *http.Request) error {
+	d := request.FormValue("delay")
+	if len(d) == 0 {
+		return nil
+	}
+
+	t, err := time.ParseDuration(d)
+	if err != nil {
+		return err
+	}
+	time.Sleep(t)
+	return nil
+}
+
+func echoHandler(w http.ResponseWriter, r *http.Request) {
+	fmt.Printf("Echoing back request made to %s to client (%s)\n", r.RequestURI, r.RemoteAddr)
+
+	// If the request has form ?delay=[:duration] wait for duration
+	// For example, ?delay=10s will cause the response to wait 10s before responding
+	if err := delayResponse(r); err != nil {
```

Comment:
I wonder if we need to test this? Maybe just a simple unit test that calls `delayRepsonse` with a very short sleep value, just to make sure the query form parsing works? Or is this simple enough that we can assume it works, given that it's only used (and tested by) the timeouts conformance test? @robscott WDYT?

Reply:
@frankbu I only tested this manually and it seems to work fine. I can add unit tests if required but I thought that the conformance test would fail if this is not working as expected. Maybe my assumption was too frail? 

![Screenshot 2023-10-02 at 16 38 56](https://github.com/kubernetes-sigs/gateway-api/assets/32323373/0ed3fb61-ad95-427f-b37f-42de41e3b3a2)

Reply:
I think it could be useful to have some basic unit tests for our echo server, but I don't think that needs to block this PR.

Reply:
I agree, logging an issue to make sure we don't forget this can close this one out for now, I think.

Reply:
@robscott @youngnick I have added some basic unit tests in this commit a97ee54e3cffa70e3635952017a939ba773039a0 

Will I add a bash script and an instruction in Makefile to run these unit tests before the following ? https://github.com/kubernetes-sigs/gateway-api/blob/d38e6c81ba3d3f137012ceeec8262ab40ad0cd81/Makefile#L155

Reply:
If you could please add `./conformance/echoserver/...` to the `test:` target, that would be great.

Reply:
done @youngnick, thanks for the pointer! 

I have verified tests are running as expected when running `make`:

```
ok  	sigs.k8s.io/gateway-api/conformance/echo-basic	1.635s	coverage: 46.5% of statements
```